File: tests/payloads/suite/cpp/cpp/rabinkarp.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    long long length;
    string pattern;
    if (!(cin >> length >> pattern)) return 0;

    string text = "";
    for (int i = 0; i < length; ++i) {
        text += pattern;
    }

    long long base = 31;
    long long mod = 1e9 + 7;
    long long m = pattern.length();
    long long n = text.length();

    long long ph = 0, th = 0, power = 1;
    for (int i = 0; i < m; ++i) {
        ph = (ph * base + (long long)pattern[i]) % mod;
        th = (th * base + (long long)text[i]) % mod;
        if (i > 0) power = (power * base) % mod;
    }

    long long count = 0;
    for (int i = 0; i <= n - m; ++i) {
        if (ph == th) count++;
        if (i + m < n) {
            th = (th * base - (long long)text[i] * power * base + (long long)text[i + m]) % mod;
            if (th < 0) th += mod;
        }
    }

    cout << count << endl;

    return 0;
}

File: tests/payloads/suite/java/rabinkarp.java
import java.util.Scanner;

public class Solution {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int length = sc.nextInt();
        String pattern = sc.next();
        
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < length; i++) {
            sb.append(pattern);
        }
        String text = sb.toString();
        
        long base = 31;
        long mod = 1_000_000_007L;
        int m = pattern.length();
        int n = text.length();
        
        long ph = 0;
        long th = 0;
        long power = 1;
        
        for (int i = 0; i < m; i++) {
            ph = (ph * base + (int) pattern.charAt(i)) % mod;
            th = (th * base + (int) text.charAt(i)) % mod;
            if (i > 0) {
                power = (power * base) % mod;
            }
        }
        
        int count = 0;
        for (int i = 0; i <= n - m; i++) {
            if (ph == th) {
                count++;
            }
            if (i + m < n) {
                long term1 = (th * base) % mod;
                long term2 = (((long) text.charAt(i) * power) % mod * base) % mod;
                th = (term1 - term2 + (int) text.charAt(i + m)) % mod;
                if (th < 0) {
                    th += mod;
                }
            }
        }
        System.out.println(count);
    }
}
